File: built-in plugins dsp/mb-pad-dark.cpp
/**
 * MB Dark Pad
 * Category : instrument
 * Type     : pad
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Moody dark atmosphere
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_PAD_DARK_H
#define MB_PAD_DARK_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbPadDark : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-pad-dark";
    static constexpr const char* PLUGIN_NAME    = "MB Dark Pad";
    static constexpr const char* PLUGIN_TYPE    = "pad";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float darkness = 0.8f;  // range [0, 1]
    float volume = 0.7f;  // range [0, 1]
    };

    MbPadDark() = default;
    ~MbPadDark() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.darkness = std::clamp(params.darkness, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Dark Pad
        return input;
    }
};

#endif // MB_PAD_DARK_H

File: built-in plugins dsp/mb-pad-dark.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbPadDarkPlugin: PluginDefinition = {
  id: "mb-pad-dark",
  slug: "mb-pad-dark",
  name: "MB Dark Pad",
  category: "instrument",
  type: "pad",
  version: "1.0.0",
  description: "Moody dark atmosphere",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sawtooth", detune: -12, gain: 0.3 },
    { type: "sawtooth", detune: 12, gain: 0.3 },
    { type: "sine", detune: -1200, gain: 0.4 },
  ],
  envelope: { attack: 1.5, decay: 1.0, sustain: 0.8, release: 3.0 },
  parameters: [
    {
      id: "darkness",
      name: "Darkness",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { darkness: 0.8, volume: 0.7 },
};

export default MbPadDarkPlugin;
